include/sparrow/types/arrow_length.hpp

```diff
@@ -0,0 +1,166 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <limits>
+#include <cstdint>
+#include <concepts>
+#include <stdexcept>
+#include <utility>
+#include <string>
+#include <typeindex>
+
+#include "sparrow/config/config.hpp"
+
+namespace sparrow
+{
+
+    /// Type representing an Arrow length (sizes, offsets) in the Arrow specification and storage.
+    /// This is used internally but is not always directly convertible to `std::size_t`
+    /// and `std::ptrdiff_t` which are the "native" types to represent sizes and offsets.
+    /// For conversion purposes:
+    /// - @see `to_native_size()`
+    /// - @see `to_native_offset()`
+    /// - @see `to_arrow_length()`
+    /// - @see `sum_arrow_offsets()`
+    using arrow_length = std::int64_t;
+
+    /// Clarifies if a length value is supposed to be a size/length or an offset.
+    /// Offsets can be negative, sizes cannot. This is only important for runtime checks
```

Comment:
in which context can offsets be negative?

Reply:
They cannot, I removed `arrow_length_kind`

Reply:
Offsets in C++ standard can be negative, this is used to find a new index given a position or to compare positions and obtain a potentially negative (relative to the first element) offset. That's why `std::ptrdiff_t` is signed. 
Note that this is about the C++ standard side of things, not the Arrow side.
I didnt initially needed this distinction until I encountered lots of code in sparrow that was mixing `ptrdiff_t` and `size_t` to compute a position.